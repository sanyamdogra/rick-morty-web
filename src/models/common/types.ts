import { CharacterResponse, LocationResponse } from "../../services/types";

export interface NameLoc {
  name: string;
  url: string;
}

export interface Character {
  created: string;
  episode: string[];
  gender: string;
  id: number;
  image: string;
  location: NameLoc;
  name: string;
  origin: NameLoc;
  species: string;
  status: string;
  type: string;
  url: string;
}

export interface Episode {
  air_date: string;
  characters: string[];
  created: string;
  episode: string;
  id: number;
  name: string;
  url: string;
}

export type MockData =
  | Character
  | CharacterResponse
  | Episode
  | LocationResponse;
