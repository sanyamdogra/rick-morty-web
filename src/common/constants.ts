export const GITHUB_PROJECT_REPO =
  "https://github.com/sanyamdogra/rick-morty-web";
