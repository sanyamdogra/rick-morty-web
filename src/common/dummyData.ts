import { Character, Episode } from "../models/common/types";
import { CharacterResponse, LocationResponse } from "../services/types";

export const DUMMY_CHARACTER: Character = {
  created: "2017-11-04T18:48:46.250Z",
  episode: [
    "https://rickandmortyapi.com/api/episode/1",
    "https://rickandmortyapi.com/api/episode/2",
  ],
  gender: "Male",
  id: 1,
  image: "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
  location: {
    name: "Citadel of Ricks",
    url: "https://rickandmortyapi.com/api/location/3",
  },
  name: "Rick Sanchez",
  origin: {
    name: "Earth (C-137)",
    url: "https://rickandmortyapi.com/api/location/1",
  },
  species: "Human",
  status: "Alive",
  type: "",
  url: "https://rickandmortyapi.com/api/character/1",
};

export const DUMMY_EPISODES: Episode[] = [
  {
    air_date: "December 2, 2013",
    characters: [
      "https://rickandmortyapi.com/api/character/1",
      "https://rickandmortyapi.com/api/character/2",
    ],
    created: "2017-11-10T12:56:33.798Z",
    episode: "S01E01",
    id: 1,
    name: "Pilot",
    url: "https://rickandmortyapi.com/api/episode/1",
  },
  {
    air_date: "December 2, 2013",
    characters: [
      "https://rickandmortyapi.com/api/character/1",
      "https://rickandmortyapi.com/api/character/2",
    ],
    created: "2017-11-10T12:56:33.798Z",
    episode: "S01E01",
    id: 1,
    name: "Pilot",
    url: "https://rickandmortyapi.com/api/episode/1",
  },
];

export const DUMMY_EPISODE: Episode = {
  air_date: "December 2, 2013",
  characters: [
    "https://rickandmortyapi.com/api/character/1",
    "https://rickandmortyapi.com/api/character/2",
  ],
  created: "2017-11-10T12:56:33.798Z",
  episode: "S01E01",
  id: 1,
  name: "Pilot",
  url: "https://rickandmortyapi.com/api/episode/1",
};

export const DUMMY_LOCATION: LocationResponse = {
  created: "2017-11-10T13:08:13.191Z",
  dimension: "unknown",
  id: 3,
  name: "Citadel of Ricks",
  residents: [
    "https://rickandmortyapi.com/api/character/8",
    "https://rickandmortyapi.com/api/character/14",
  ],
  type: "Space station",
  url: "https://rickandmortyapi.com/api/location/3",
};

export const DUMMY_CHARACTER_RESPONSE: CharacterResponse = {
  info: {
    count: 826,
    next: "https://rickandmortyapi.com/api/character?page=2",
    pages: 42,
    prev: null,
  },
  results: [DUMMY_CHARACTER],
};
