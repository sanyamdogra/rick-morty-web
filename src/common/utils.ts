import { MockData } from "../models/common/types";

export const getEpisodeParams = (arr: string[]) => {
  return arr.reduce((acc: number[], str: string) => {
    const strArr = str.split("/");
    acc.push(parseInt(strArr[strArr.length - 1]));

    return acc;
  }, []);
};

export const mockFetchResponse = (data: MockData) => {
  global.fetch = jest.fn().mockImplementationOnce(() =>
    Promise.resolve({
      json: () => Promise.resolve(data),
    })
  );
};
