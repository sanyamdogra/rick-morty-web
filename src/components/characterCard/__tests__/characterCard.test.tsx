import React from "react";
import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import CharacterCard from "..";

import { DUMMY_CHARACTER } from "../../../common/dummyData";

afterEach(cleanup);

test("should match the snapshot and render correct data", () => {
  const { asFragment } = render(<CharacterCard character={DUMMY_CHARACTER} />);

  expect(asFragment()).toMatchSnapshot();

  expect(screen.getByTestId("character-card-name")).toHaveTextContent(
    "Rick Sanchez"
  );

  expect(screen.getByTestId("character-card-species")).toHaveTextContent(
    "Human"
  );

  const image = screen.getByTestId("character-card-image");
  expect(image).toHaveAttribute(
    "src",
    "https://rickandmortyapi.com/api/character/avatar/1.jpeg"
  );
  expect(image).toHaveAttribute("alt", "Rick Sanchez");
});

test("clicking on the last location should open/close episode modal", () => {
  render(<CharacterCard character={DUMMY_CHARACTER} />);

  fireEvent.click(screen.getByText("Citadel of Ricks"));

  expect(screen.getByTestId("modal-body")).toBeInTheDocument();

  fireEvent.click(screen.getByTestId("modal-close-btn"));

  expect(screen.queryByTestId("modal-body")).toBeNull();
});

test("clicking on the appears on should open/close location modal", () => {
  render(<CharacterCard character={DUMMY_CHARACTER} />);

  fireEvent.click(screen.getByText("2 chapters"));

  expect(screen.getByTestId("modal-body")).toBeInTheDocument();

  fireEvent.click(screen.getByTestId("modal-close-btn"));

  expect(screen.queryByTestId("modal-body")).toBeNull();
});
