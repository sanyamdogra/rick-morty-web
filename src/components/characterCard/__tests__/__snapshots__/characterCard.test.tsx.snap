// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot and render correct data 1`] = `
<DocumentFragment>
  <div
    class="rick-morty-character-card"
  >
    <div
      class="rick-morty-character-card__image-wrapper"
    >
      <img
        alt="Rick Sanchez"
        data-testid="character-card-image"
        src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
      />
    </div>
    <div
      class="rick-morty-character-card__details"
    >
      <div
        class="rick-morty-character-card__name"
        data-testid="character-card-name"
        title="Rick Sanchez"
      >
        Rick Sanchez
      </div>
      <div
        class="rick-morty-character-card__value"
        data-testid="character-card-species"
      >
        <div
          class="rick-morty-status-indicator rick-morty-status-indicator--alive"
          data-testid="status-indicator"
          title="Alive"
        />
        Human
      </div>
      <div
        class="rick-morty-card-field__label"
        data-testid="card-field-label"
      >
        Origin
      </div>
      <div
        class="rick-morty-card-field__value rick-morty-card-field__value--none"
        data-testid="card-field-value"
      >
        Earth (C-137)
      </div>
      <div
        class="rick-morty-card-field__label"
        data-testid="card-field-label"
      >
        Last location
      </div>
      <div
        class="rick-morty-card-field__value rick-morty-card-field__value--invert"
        data-testid="card-field-value"
        title="Click for more info on location"
      >
        Citadel of Ricks
      </div>
      <div
        class="rick-morty-card-field__label"
        data-testid="card-field-label"
      >
        Appears on
      </div>
      <div
        class="rick-morty-card-field__value rick-morty-card-field__value--invert"
        data-testid="card-field-value"
        title="Click for more info on episodes"
      >
        2 chapters
      </div>
    </div>
  </div>
</DocumentFragment>
`;
